### PR TITLE
feat: simplifies Feature API

### PR DESCRIPTION
**Pull request by bartoszmajsak**
## Description

- `FeaturesInitializer` (now named `FeaturesHandler`) becomes an entry point to compose features 
    - it has two modes: cluster and component. Based on this mode it will initialize `.spec.source` transparently 
    - there is no `Prepare` func anymore - it is now part of `Apply` and `Delete` - most importantly `FeaturesHandler` keeps track of created features so there is no need to keep adding to the slice explicitly
- Tests have been adjusted to focus on behaviour rather than implementation details

## How Has This Been Tested?

- k8s testenv tests has been adjusted
- manually tested - creating DSCI and DSC with KServe. ✅ setting KServe to `Removed` cleans up knative-serving as before

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: tests/integration/features/features_int_test.go [hunk 1]
```diff
@@ -16,10 +16,10 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
-	dscv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/dscinitialization/v1"
+	dsciv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/dscinitialization/v1"
```
Comment (cam-garrison) [inline tests/integration/features/features_int_test.go:19]: good catch :smile:

**Review by cam-garrison — APPROVED** [review]
I like the brevity this brings :+1: plus keeping things behind ComponentFeaturesHandler + ClusterFeaturesHandler exposes less (ie the source).

File: pkg/feature/builder.go [hunk 2]
```diff
@@ -13,58 +13,48 @@ import (
...
-	}
-}
-
-type featureSource struct {
-	spec *v1.DSCInitializationSpec
-	name string
-}
-
-func (fo *featureSource) From(source featurev1.Source) *featureBuilder {
+func (u *usingFeaturesHandler) For(featuresHandler *FeaturesHandler) *featureBuilder {
```
Comment (israel-hdez) [inline pkg/feature/builder.go:40]: `For` is reborn!

**Review by israel-hdez — APPROVED** [review]
Yes, definitely an improvement, since implementation details are less exposed.

The change to the tests, although fine, I'm still wondering if this is a good pattern. Both on the before and after my though is that tests aren't testing with precision what is being running, since we need to build a feature from scratch, rather than using the real ones. I can guess the current approach is a trade-off, since testing the real features would mean a large amount of assertions to check in a single unit.

File: pkg/feature/builder.go (hunk 2 above)
Comment (bartoszmajsak) [reply pkg/feature/builder.go:40]: `OwnedBy`, `HandledBy`... naming is hard. The idea here is that handler will have an ownership and control over this feature. Somewhat tightly coupled using the builder, but OTOH this makes it easier to... handle... on the consuming side (controllers)

**Comment by bartoszmajsak** [comment]
> Both on the before and after my though is that tests aren't testing with precision what is being running, since we need to build a feature from scratch, rather than using the real ones. I can guess the current approach is a trade-off, since testing the real features would mean a large amount of assertions to check in a single unit.

@israel-hdez I am also not really happy with the current approach and I have it in the back of my head. It's ok to test some primitives like we do know - e.g. preconditions, processing templates and so, but it is not really covering the real behaviour of the operator itself. There's this background thread in my head processing this, I hope it will soon finish execution and I will have something better to share.

File: pkg/feature/manifest.go [hunk 3]
```diff
@@ -14,6 +14,12 @@ import (
 //go:embed templates
 var embeddedFiles embed.FS
 
+const (
+	BaseDir        = "templates"
+	ServiceMeshDir = BaseDir + string(filepath.Separator) + "servicemesh"
```
Comment (zdtsw) [inline pkg/feature/manifest.go:19]:
```suggestion
	ServiceMeshDir = path.Join(BaseDir, "servicemesh")
```

Comment (bartoszmajsak) [reply pkg/feature/manifest.go]: Yes... I am tired :) Thank you, fixed in d873c21

**Review by zdtsw — APPROVED** [review]

**Pull request merged**